File: python/semanticabi/steps/ParameterFlattener.py
from abc import ABC, abstractmethod
from typing import List

from semanticabi.abi.item.SemanticAbiItem import SemanticAbiItem
from semanticabi.abi.item.SemanticParameter import SemanticParameters, SemanticParameter
from semanticabi.common.column.DatasetColumn import DatasetColumn
from semanticabi.steps.FlattenedParameter import FlattenedParameter


class FlattenPredicate(ABC):
    """
    Predicate to determine what parameters to flatten
    """

    @abstractmethod
    def should_flatten(self, parameter: SemanticParameter, path: List[SemanticParameter]) -> bool:
        """
        Whether to flatten the components of the parameter
        """
        pass


class DefaultFlattenPredicate(FlattenPredicate):
    """
    Default predicate that flattens all non-array parameters
    """
    def should_flatten(self, parameter: SemanticParameter, path: List[SemanticParameter]) -> bool:
        """
        Don't flatten any array parameters
        """
        return not parameter.parameter.is_array


class ParameterFlattener:
    """
    Iterates through the parameters of an abi, flattening any tuple components to individual columns, and returning
    a list of the parameters that can be used to build the schema.

    For primitive parameters, we just create a column with the parameter name as the column name. For tuple parameters,
    we recursively go through the components, and create a column for each primitive component with a name that
    combines the parent parameter name and the individual parameter's name to ensure uniqueness.
    """
    _abi_item: SemanticAbiItem
    _predicate: FlattenPredicate
    _flattened_parameters: List[FlattenedParameter]

    @staticmethod
    def build_column_name(path: List[SemanticParameter], name: str) -> str:
        """
        Create a flattened column name for a parameter that is nested in a tuple
        """
        return '_'.join([p.name for p in path] + [name])

    def __init__(self, abi_item: SemanticAbiItem, predicate: FlattenPredicate = DefaultFlattenPredicate()):
        self._abi_item = abi_item
        self._predicate = predicate
        self._flattened_parameters = self._build_parameter_list()

    def parameter_list(self) -> List[FlattenedParameter]:
        """
        Get the list of flattened parameters for this abi item
        """
        return self._flattened_parameters

    def dataset_columns(self) -> List[DatasetColumn]:
        """
        Get the list of dataset columns for this abi item
        """
        return [parameter.final_dataset_column for parameter in self._flattened_parameters]

    def _build_parameter_list(self) -> List[FlattenedParameter]:
        parameters = self._flatten_parameters(self._abi_item.input_parameters, True)

        if self._abi_item.output_parameters is not None:
            parameters += self._flatten_parameters(self._abi_item.output_parameters, False)

        return parameters

    def _flatten_parameters(
        self,
        semantic_parameters: SemanticParameters,
        is_input: bool,
        path: List[SemanticParameter] = None
    ) -> List[FlattenedParameter]:
        if path is None:
            path = []

        flattened_parameters: List[FlattenedParameter] = []
        for semantic_trace_parameter in semantic_parameters.parameters().values():
            if semantic_trace_parameter.exclude:
                continue

            if not self._predicate.should_flatten(semantic_trace_parameter, path):
                continue

            flattened_parameters += self._flatten_parameter(semantic_trace_parameter, is_input, path)

        return flattened_parameters

    def _flatten_parameter(
        self,
        semantic_parameter: SemanticParameter,
        is_input: bool,
        path: List[SemanticParameter]
    ) -> List[FlattenedParameter]:
        if semantic_parameter.components is not None:
            # If we have a tuple, recursively flatten the components
            return self._flatten_parameters(
                semantic_parameter.components,
                is_input,
                path + [semantic_parameter]
            )
        else:
            return [FlattenedParameter(
                semantic_parameter,
                path,
                ParameterFlattener.build_column_name(path, semantic_parameter.name),
                is_input
            )]

File: js/src/semanticabi/transform/steps/ParameterFlattener.ts
import {SemanticAbiItem} from "semanticabi/abi/item/semantic/SemanticAbiItem";
import {SemanticParameter, SemanticParameters} from "semanticabi/abi/item/semantic/SemanticParameters";
import {FlattenedParameter} from "semanticabi/transform/steps/FlattenedParameter";
import {SemanticAbiFunction} from "semanticabi/abi/item/semantic/SemanticAbiFunction";
import {Optional} from "common/Optional";
import {DatasetColumn} from "semanticabi/transform/column/DatasetColumn";

/**
 * Iterates through the parameters of an abi, flattening any tuple components to individual columns, and returning a
 * list of the parameters that can be used to build the schema.
 *
 * For primitive parameters, we just create a column with the parameter name as the column name. For tuple parameters,
 * we recursively go through the components, and create a column for each primitive component with a name that combines
 * the parent parameter name and the individual parameter's name to ensure uniqueness.
 */
export class ParameterFlattener {

    public readonly flattened: FlattenedParameter[];

    constructor(
        private readonly abiItem: SemanticAbiItem,
        private readonly predicate: FlattenPredicate = new FlattenAllPredicate()
    ) {
        this.flattened = ParameterFlattener.buildParameterList(this.abiItem, this.predicate);
    }

    get datasetColumns(): DatasetColumn[] {
        return this.flattened.map(c => c.finalDatasetColumn);
    }

    public static buildParameterList(abiItem: SemanticAbiItem, predicate: FlattenPredicate): FlattenedParameter[] {
        return [
            ...ParameterFlattener.flattenParameters(abiItem.inputParameters, true, predicate),
            ...Optional.ofType(abiItem, SemanticAbiFunction)
                .map(c => ParameterFlattener.flattenParameters(c.outputParameters, false, predicate))
                .getOr([])
        ];
    }

    private static flattenParameters(
        semanticParameters: SemanticParameters,
        isInput: boolean,
        predicate: FlattenPredicate,
        path: SemanticParameter[] = []
    ): FlattenedParameter[] {
        return Array.from(semanticParameters.parameters.values())
            .filter(parameter => !parameter.exclude && predicate.shouldFlatten(parameter, path))
            .flatMap(parameter => ParameterFlattener.flattenParameter(parameter, isInput, predicate, path));
    }

    private static flattenParameter(
        semanticParameter: SemanticParameter,
        isInput: boolean,
        predicate: FlattenPredicate,
        path: SemanticParameter[]
    ): FlattenedParameter[] {
        return semanticParameter.components
            // Recursively flatten the components of a tuple.
            .map(parameters => ParameterFlattener.flattenParameters(parameters, isInput, predicate, path.concat(semanticParameter)))
            // Flatten a single parameter.
            .getOrElse(() => [new FlattenedParameter(
                semanticParameter,
                path,
                ParameterFlattener.buildColumnName(path, semanticParameter.name),
                isInput
            )]);
    }

    private static buildColumnName(path: SemanticParameter[], name: string): string {
        return [...path.map(p => p.name), name].join('_');
    }

}

/**
 * Predicate to determine what parameters to flatten.
 */
export interface FlattenPredicate {

    shouldFlatten(parameter: SemanticParameter, path: SemanticParameter[]): boolean;

}

/**
 * Flattens all non-array parameters.
 */
export class FlattenAllPredicate implements FlattenPredicate {

    shouldFlatten(parameter: SemanticParameter, path: SemanticParameter[]): boolean {
        return !parameter.parameter.isArray;
    }

}